### ISSUE TITLE
Question about finetuning on the LeRobot dataset?

### ISSUE THREAD

**Issue by swords123** [X-Square-Robot/wall-x#26]
Hello, I encountered an issue while reading the code. In the PreprocessedDataset class of the wall_x/data/load_lerobot_dataset.py file, the defined state space is a 20-dimensional array with each dimension defined as: ['follow_left_ee_cartesian_pos', 'follow_left_ee_rotation', 'follow_left_gripper', 'follow_right_ee_cartesian_pos', 'follow_right_ee_rotation', 'follow_right_gripper', 'head_actions', 'height', 'car_pose']. Here, the robot arm state uses the 6-degree-of-freedom translation and rotation of the end effector.
However, in the LeRobotDataset class of the lerobot/src/lerobot/datasets/lerobot_dataset.py file, the __getitem__() function returns a 14-dimensional array of 'observation.state', where each dimension corresponds to: ['left_waist', 'left_shoulder', 'left_elbow', 'left_forearm_roll', 'left_wrist_angle', 'left_wrist_rotate', 'left_gripper', 'right_waist', 'right_shoulder', 'right_elbow', 'right_forearm_roll', 'right_wrist_angle', 'right_wrist_rotate', 'right_gripper']. This uses the robot arm's joint angles to represent the robot's state space.
There is no forward kinematics code in the codebase, and the joint angles are directly used as input to the model's state space for fine-tuning on the LeRobot dataset. So, I would like to ask:
1. During actual training, do you use the robot arm's end effector or joint angles as the model input?
2. When fine-tuning on LeRobot, can good results be achieved by directly using joint angles as input?

你好，在阅读代码过程中我遇到一个问题。在wall_x/data/load_lerobot_dataset.py文件的PreprocessedDataset类中，定义的状态空间是一个20维的数组，各个维度定义为：[ 'follow_left_ee_cartesian_pos', 'follow_left_ee_rotation'，'follow_left_gripper','follow_right_ee_cartesian_pos','follow_right_ee_rotation','follow_right_gripper','head_actions','height', 'car_pose']， 其中关于机械臂状态使用的是末端执行器6自由度的平移和旋转量。而在lerobot/src/lerobot/datasets/lerobot_dataset.py文件的LeRobotDataset类中，__getitem__()函数返回的是一个14维的数组，且这14维中每个维度的含义是[ 'left_waist','left_shoulder','left_elbow','left_forearm_roll', 'left_wrist_angle', 'left_wrist_rotate', 'left_gripper', 'right_waist', 'right_shoulder', 'right_elbow', 'right_forearm_roll', 'right_wrist_angle', 'right_wrist_rotate','right_gripper'], 其使用的机械臂关节角度表示的机器人的状态空间。代码中没有任何前向运动学的代码，而是直接将关节角度作为模型的状态空间作为输入。我想问下你们在实际训练过程中，到底用的机械臂的末端还是机械臂的关节作为模型输入？以及在LeRobot上微调时，直接将关节角度作为输入可以得到不错的结果吗？

**Comment by liangyuxin42** [comment]
Hi, thank you for your detailed analysis and for raising this excellent question.
To clarify, during the pre-training phase of our model, we use an action space based on the absolute pose of the end-effector.
However, when fine-tuning on other datasets like LeRobot or adapting to different robot embodiments, the action space can indeed be changed. For instance, we have successfully fine-tuned models on SO-101 using joint-based action spaces and achieved expected results.
When you switch the action space, the crucial step is to reconfigure the dof_config and adjust the action normalization statistics to match the new action representation. [code ref](https://github.com/X-Square-Robot/wall-x/blob/4fa32fb18803959da24a674ab2d74d7778efd610/wall_x/model/qwen2_5_based/modeling_qwen2_5_vl_act.py#L750) [Explanation](https://github.com/X-Square-Robot/wall-x/blob/main/workspace/README.md#customize-your-robot-configuration)
Hope this helps!
